File: leetcode/242. Valid Anagram/solution.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length !== t.length) return false;

    const count = new Array(26).fill(0);

    const ascii_value_of_a = 'a'.charCodeAt(0);

    for (let i = 0; i < s.length; i++) {
        count[s.charCodeAt(i) - ascii_value_of_a]++;
        count[t.charCodeAt(i) - ascii_value_of_a]--;
    }

    return count.every(c => c === 0);
};

File: leetcode/242. Valid Anagram/solution.c
#include <stdbool.h>
#include <string.h>

// Time: O(3n + 26) ~> O(n)
// Space: O(26) ~> O(1)
bool isAnagram(char* s, char* t) {
    int len = strlen(s);
    if (len != strlen(t)) return false;

    int count[26] = {};

    for (int i = 0; i < len; i++) {
        count[s[i] - 'a']++;
        count[t[i] - 'a']--;
    }
    for (int i = 0; i < 26; i++)
        if (count[i] != 0) return false;

    return true;
}
